### PR TITLE
Integrate Erlang's metadata and Elixir's metadata

### PR DESCRIPTION
**Pull request by hauleth**
Close erlef/eef-observability-wg#6

### CODE REVIEW

**Comment by fertapric** [comment]
Elixir's Logger metadata allows to have duplicated keys, so this would be a backwards incompatible change. Said that, I'm not sure how common that is 🙂 

Thoughts?

**Comment by josevalim** [comment]
Hi @hauleth, thanks for the PR! This also requires Erlang/OTP 21, which means we can’t merge it any sooner than the other PR. So maybe we should get all of the work done at once in that PR?

**Comment by josevalim** [comment]
And to clarify, I don’t want to do things conditionally, because it may lead to subtle difference of behavior as mentioned by @fertapric. If there is a need to rush, I would rather merge the complete PR and drop support for 20 sooner.

File: lib/logger/lib/logger.ex [hunk 1]
```diff
@@ -447,53 +447,50 @@ defmodule Logger do
...
 
-  defp into_metadata([], metadata), do: metadata
-  defp into_metadata(keyword, metadata), do: into_metadata(keyword, [], metadata)
-
-  defp into_metadata([{key, nil} | keyword], prepend, metadata) do
-    into_metadata(keyword, prepend, :lists.keydelete(key, 1, metadata))
-  end
+      map when is_map(map) ->
+        merged = Enum.into(keyword, map)
+        metadata = filter_out_nils(merged)
```
Comment (michalmuskala) [inline lib/logger/lib/logger.ex]: Is the intention to also filter the existing metadata or just the new ones? If only the new one, we could possibly merge & filter in one go
```elixir
for {_k, v} = elem <- keyword, v != nil, into: map, do: elem
```

**Comment by hauleth** [comment]
@fertapric while in theory having duplicated keys in metadata is possible, it is highly impropable, as almost any action on these will remove duplicates anyway.

File: lib/logger/lib/logger.ex (hunk 1 above)
Comment (hauleth) [reply lib/logger/lib/logger.ex]: @michalmuskala when sending data to the backend all nils will be removed anyway, so there is not much difference.

**Comment by hauleth** [comment]
@josevalim I proposed it as a result of discussion during WG meeting where there was proposal to provide metadata integration sooner, before whole #9333 will be ready to merge. But in the end it is up to Elixir team how to handle that.

**Comment by josevalim** [comment]
We have three months if we bring the release date early and nine months otherwise, so let's try to make it go the whole way in. :)

**Comment by josevalim** [comment]
Thanks for the PR!

**Pull request closed without merging**